File: PROJECTS/encrypted-p2p-chat/frontend/src/types/websocket.ts
// ===================
// © AngelaMos | 2025
// websockets.ts
// ===================
import type { PresenceStatus } from "./chat"

export type WSMessageType =
  | "encrypted_message"
  | "typing"
  | "presence"
  | "receipt"
  | "heartbeat"
  | "error"

export interface BaseWSMessage {
  type: WSMessageType
  timestamp?: string
}

export interface EncryptedMessageWS extends BaseWSMessage {
  type: "encrypted_message"
  message_id: string
  sender_id: string
  recipient_id: string
  room_id: string
  ciphertext: string
  nonce: string
  header: string
  sender_username: string
}

export interface TypingIndicatorWS extends BaseWSMessage {
  type: "typing"
  user_id: string
  room_id: string
  is_typing: boolean
  username: string
}

export interface PresenceUpdateWS extends BaseWSMessage {
  type: "presence"
  user_id: string
  status: PresenceStatus
  last_seen: string
}

export interface ReadReceiptWS extends BaseWSMessage {
  type: "receipt"
  message_id: string
  room_id: string
  user_id: string
  read_at: string
}

export interface HeartbeatWS extends BaseWSMessage {
  type: "heartbeat"
}

export interface ErrorMessageWS extends BaseWSMessage {
  type: "error"
  error_code: string
  error_message: string
  details?: Record<string, unknown>
}

export type WSMessage =
  | EncryptedMessageWS
  | TypingIndicatorWS
  | PresenceUpdateWS
  | ReadReceiptWS
  | HeartbeatWS
  | ErrorMessageWS

export interface WSOutgoingEncryptedMessage {
  type: "encrypted_message"
  recipient_id: string
  room_id: string
  plaintext: string
}

export interface WSOutgoingTyping {
  type: "typing"
  room_id: string
  is_typing: boolean
}

export interface WSOutgoingPresence {
  type: "presence"
  status: PresenceStatus
}

export interface WSOutgoingReceipt {
  type: "receipt"
  message_id: string
  room_id: string
}

export interface WSOutgoingHeartbeat {
  type: "heartbeat"
  timestamp: string
}

export type WSOutgoingMessage =
  | WSOutgoingEncryptedMessage
  | WSOutgoingTyping
  | WSOutgoingPresence
  | WSOutgoingReceipt
  | WSOutgoingHeartbeat

export const WS_HEARTBEAT_INTERVAL = 30000
export const WS_RECONNECT_DELAY = 5000
export const WS_MAX_RECONNECT_ATTEMPTS = 10

File: PROJECTS/encrypted-p2p-chat/backend/app/schemas/websocket.py
"""
ⒸAngelaMos | 2025
Pydantic schemas for WebSocket message types
"""

from typing import Any
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.config import (
    ENCRYPTED_CONTENT_MAX_LENGTH,
    MESSAGE_ID_MAX_LENGTH,
)


class BaseWSMessage(BaseModel):
    """
    Base WebSocket message with common fields
    """
    type: str
    timestamp: datetime | None = None


class EncryptedMessageWS(BaseWSMessage):
    """
    Encrypted message sent over WebSocket
    """
    type: str = "encrypted_message"
    message_id: str = Field(max_length = MESSAGE_ID_MAX_LENGTH)
    sender_id: str
    recipient_id: str
    ciphertext: str = Field(max_length = ENCRYPTED_CONTENT_MAX_LENGTH)
    nonce: str
    header: str
    sender_username: str


class TypingIndicatorWS(BaseWSMessage):
    """
    Typing indicator message
    """
    type: str = "typing"
    user_id: str
    room_id: str
    is_typing: bool


class PresenceUpdateWS(BaseWSMessage):
    """
    User presence update message
    """
    type: str = "presence"
    user_id: str
    status: str
    last_seen: datetime


class ReadReceiptWS(BaseWSMessage):
    """
    Message read receipt
    """
    type: str = "receipt"
    message_id: str = Field(max_length = MESSAGE_ID_MAX_LENGTH)
    user_id: str
    read_at: datetime


class ErrorMessageWS(BaseWSMessage):
    """
    Error message sent over WebSocket
    """
    type: str = "error"
    error_code: str
    error_message: str
    details: dict[str, Any] | None = None


class WSConnectionRequest(BaseModel):
    """
    WebSocket connection request with auth token
    """
    user_id: UUID
    token: str | None = None


class WSHeartbeat(BaseModel):
    """
    WebSocket heartbeat ping/pong message
    """
    type: str = "heartbeat"
    timestamp: datetime
